### ISSUE TITLE
Attempting to log metrics per epoch returns a "step must increase" warning.

### ISSUE THREAD

**Issue by seanswyi** [wandb/client#1591]
**Describe the bug**
Apologies if the title is a bit ambiguous. I'm also not entirely sure if this is a "bug," but am hoping that someone can help out. Currently I'm training and evaluating my model with two loops, one outer loop for epochs and one inner loop for the batched data.

I want to log separate plots for my metrics _per epoch_ (e.g., one line plot for epoch 1, another for epoch 2, etc.) and I set the `step` keyword argument in `wandb.log` to the `step` in my enumerate statement. However, doing so gave me a warning:

```
wandb: WARNING Step must only increase in log calls.  Step x < y; dropping {'loss': something}.
```

I'm assuming this is because `step` gets reset to `0` every loop, and also that each plot is only meant to refer to one W&B run. I'm wondering if I can bypass that behavior and make it so that each epoch would refer to one plot. I've taken a look a [the documentation for Incremental Logging](https://docs.wandb.com/library/log#incremental-logging) but my problem wasn't solved.

**To Reproduce**
You can run the following script as is and should get the problem.

```
import torch
import torch.nn as nn
import torch.optim as optim
import wandb


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.linear1 = nn.Linear(in_features=400, out_features=100)
        self.output_linear = nn.Linear(in_features=100, out_features=10)

    def forward(self, x):
        output1 = self.linear1(x)
        output2 = self.output_linear(output1)

        return output2


def main():
    model = Model()
    input_data = torch.randn(size=(3, 300, 400))
    targets = torch.empty(300, dtype=torch.long).random_(3)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(params=model.parameters())

    for epoch in range(3):
        for step, batch in enumerate(input_data):
            optimizer.zero_grad()
            output = model(batch)
            loss = criterion(output, targets)
            loss.backward()
            optimizer.step()

            # Removing the step assignment eliminates the warning.
            wandb.log({'loss': loss.item()}, step=step)


if __name__ == '__main__':
    wandb.init(project='testing-wandb')
    main()
```

**Operating System**
 - OS: Ubuntu 16.04
 - Browser: None.
 - Version: 0.10.12

**Comment by cvphelps** [comment]
Hi @seanswyi, thanks for reaching out. We recommend not setting step here, because it needs to always increase. Why do you want a separate chart for each epoch? Most of our users plot a time series and have this all on a single chart.

If you want different metrics for each epoch, name your metrics loss_epoch_1, loss_epoch_2, etc. 

If you want to plot your metrics against different x-axes, you can log the step as a metric, like `wandb.log({'loss': 0.1, 'epoch': 1, 'batch': 3})`. In the UI you can switch between x-axes in the chart settings.

**Comment by tyomhak** [comment]
Hey @seanswyi  , thanks for writing in! 
The step you are setting in wandb.log is used internally, to track the history of the logs done. This means that each consecutive step must not be smaller than the previous one. 

For your case I would suggest using a separate metric (lets call it 'custom_step') for tracking the step, and pass it as a common metric to the wandb.log function. Something like this should work for you:

`
for epoch in range(3):
        for custom_step, batch in enumerate(input_data):
            optimizer.zero_grad()
            output = model(batch)
            loss = criterion(output, targets)
            loss.backward()
            
            wandb.log({'loss': loss.item(), 'custom_step':custom_step})
`

After this, you could change the x-axis metric for the charts from 'step' to 'custom_step' in your workspace UI.

**Comment by seanswyi** [comment]
@cvphelps @tyomhak Thanks for the quick responses guys. I guess the reason why I wanted to plot separately for each epoch is because I wanted to compare how the metrics change from epoch to epoch by overlaying the plots on top of each other. Plotting the metrics as one large time series made it a little more difficult to do this, so I was wondering if there was an internal way I could achieve that.

I've also tried logging the step itself as a metric and converting the logged chart to have that logged step as the x-axis. This does allow for better presentation, but it didn't seem to give me the "laying on top of each other" view that I wanted. I suppose it's good enough though.

Thanks again. I'll close this Issue, I believe it's solved.

**Comment by cvphelps** [comment]
Thanks @seanswyi , to be clear, if you have different metric names for each — loss_1 loss_2 etc, then you can overlay those lines on a single line plot in the UI by clicking edit on a chart and adding multiple y-axes.

**Comment by Mahmoud-Khaled-Nasr** [comment]
Thanks, guys @cvphelps @tyomhak  for your responses but I am facing some trouble similar to this question and would like to ask if you have a solution for me.
I will use the same code as before. The different use case I have is that I want to log the loss per epoch, so on the y-axis is the average loss for the entire epoch and another log for the loss of each batch for the epoch. I want to do this to see the fluctuations of the loss due to the batch size.

```
for epoch in range(3):
        running_average = 0
        for custom_step, batch in enumerate(input_data):
            optimizer.zero_grad()
            output = model(batch)
            loss = criterion(output, targets)
            loss.backward()
            optimizer.step()
            running_average += loss.item()
            wandb.log({'loss': loss.item(), 'custom_step':custom_step})
         wandb.log({'epoch_loss': running_average / len(input_data), 'custom_step2':epoch})```

Is there a way to achieve this behavior?

**Comment by vanpelt** [comment]
Hey @Mahmoud-Khaled-Nasr the code you have above should achieve exactly what you need.  You'll just need to change the "x-axis" for specific charts in the UI.

**Comment by Mahmoud-Khaled-Nasr** [comment]
Oh, I have been looking wrong at this. I wanted to make the epoch the global step. I totally missed this one thanks @vanpelt

**Comment by matthieubal** [comment]
What is a powerful tool to merge git branches?
